lang/java/avro/src/main/java/org/apache/avro/file/DataFileReader.java

```diff
@@ -143,36 +143,61 @@ public void seek(long position) throws IOException {
    * {@link #next()}.
    */
   @Override
-  public void sync(long position) throws IOException {
+  public void sync(final long position) throws IOException {
     seek(position);
     // work around an issue where 1.5.4 C stored sync in metadata
-    if ((position == 0) && (getMeta("avro.sync") != null)) {
+    if ((position == 0L) && (getMeta("avro.sync") != null)) {
       initialize(sin); // re-init to skip header
       return;
     }
-    try {
-      int i = 0, b;
-      InputStream in = vin.inputStream();
-      vin.readFixed(syncBuffer);
-      do {
-        int j = 0;
-        for (; j < SYNC_SIZE; j++) {
-          if (getHeader().sync[j] != syncBuffer[(i + j) % SYNC_SIZE])
-            break;
-        }
-        if (j == SYNC_SIZE) { // matched a complete sync
-          blockStart = position + i + SYNC_SIZE;
-          return;
-        }
-        b = in.read();
-        syncBuffer[i++ % SYNC_SIZE] = (byte) b;
-      } while (b != -1);
-    } catch (EOFException e) {
-      // fall through
+
+    if (this.partialMatchTable == null) {
+      this.partialMatchTable = computePartialMatchTable(getHeader().sync);
     }
-    // if no match or EOF set start to the end position
+
+    final byte[] sync = getHeader().sync;
+    final InputStream in = vin.inputStream();
```

Comment:
Shouldn't we close this `InputStream`?

Reply:
@Fokko I don't think so.  I think the idea is to position the stream to the right place and then start reading the stream in subsequent calls.  The current implementation also behaves this way.